### PR TITLE
[WIP] Importmulti private key support

### PR DESCRIPTION
**Pull request by kallewoof**
Partially addresses #12703.
Fully addresses #9694.

This implementation simply uses the existing `importprivkey` code (moved into a separate `ImportPrivateKey` method), and does not support redeem scripts, internal, pubkeys, or keys options.

One question is if this should be simplified to where a user can put private keys in `keys` and leave `scriptPubKey` empty, or if it makes more sense to add a new field as I do here.

Another question is whether this should be incorporated more into how `importmulti` works, rather than basing it off of `importprivkey` (although with the timestamp argument, it seems like it does all we want it to).

(I was initially planning on deriving the arguments for `ProcessImport` and letting it do its thing, but realized there can be multiple scriptPubKeys for one private key, so I'd have to call `ProcessImport` multiple times, with the same private key.. which will trigger an error.)

### CODE REVIEW

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -1132,7 +1132,7 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "  ]\n"
             "2. options                 (json, optional)\n"
             "  {\n"
-            "     \"rescan\": <false>,         (boolean, optional, default: true) Stating if should rescan the blockchain after all imports\n"
+            "     \"rescan\": <bool>,          (boolean, optional, default: true) Whether a rescan of the blockchain after all imports should be executed\n"
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: In commit "Importmulti private key support",

Also fix `default=true` to be consistent with most of the code.

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -88,6 +88,38 @@ bool GetWalletAddressesForKey(CWallet * const pwallet, const CKeyID &keyid, std:
...
 
+bool ImportPrivateKey(CWallet * const pwallet, const UniValue& privkey, const std::string& label = "", const int64_t timestamp = 1)
+{
+    CKey key = DecodeSecret(privkey.get_str());
+    if (!key.IsValid()) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
+
+    CPubKey pubkey = key.GetPubKey();
+    assert(key.VerifyPubKey(pubkey));
+    CKeyID address = pubkey.GetID();
+    {
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: In commit "[wallet] [rpc] Move private key import functionality out of importprivkey",

Remove this scope?

File: src/wallet/rpcdump.cpp [hunk 3]
```diff
@@ -88,6 +88,38 @@ bool GetWalletAddressesForKey(CWallet * const pwallet, const CKeyID &keyid, std:
     return fLabelFound;
 }
 
+bool ImportPrivateKey(CWallet * const pwallet, const UniValue& privkey, const std::string& label = "", const int64_t timestamp = 1)
+{
+    CKey key = DecodeSecret(privkey.get_str());
+    if (!key.IsValid()) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: In commit "[wallet] [rpc] Move private key import functionality out of importprivkey",

Nit, move throw to new line.

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -88,6 +88,38 @@ bool GetWalletAddressesForKey(CWallet * const pwallet, const CKeyID &keyid, std:
     return fLabelFound;
 }
 
+bool ImportPrivateKey(CWallet * const pwallet, const UniValue& privkey, const std::string& label = "", const int64_t timestamp = 1)
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: In commit "[wallet] [rpc] Move private key import functionality out of importprivkey",

`static bool ...`?

File: src/wallet/rpcdump.cpp [hunk 5]
```diff
@@ -814,6 +814,20 @@ UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, const int6
...
+        if (data.exists("privkey")) {
+            const UniValue& privkey = data["privkey"];
+            if (data.exists("scriptPubKey"))           throw JSONRPCError(RPC_INVALID_PARAMETER, "Must provide either a privkey or a scriptPubKey");
+            if (strRedeemScript != "")                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Redeem script not supported for privkey import");
+            if (pubKeys.size() > 0 || keys.size() > 0) throw JSONRPCError(RPC_INVALID_PARAMETER, "Keys/pubkeys not supported for privkey import");
+            if (watchOnly)                             throw JSONRPCError(RPC_INVALID_PARAMETER, "Privkeys cannot be imported watch-only");
+            if (!ImportPrivateKey(pwallet, privkey, label, timestamp)) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the given private key");
+            }
+            UniValue result = UniValue(UniValue::VOBJ);
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: In commit "[wallet] [rpc] Add privkey support to importmulti",

Nit, `UniValue result(UniValue::VOBJ);`

**Review by promag — COMMENTED** [review]
Concept ACK. Light review.

**Comment by MeshCollider** [comment]
Concept ACK

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by sipa** [comment]
Thank you for working on this!

I think we should be careful here though. `importprivkey` has a number a quirks that are for historical or compatibility reasons, that we shouldn't maintain:
* `importprivkey` does not know what the corresponding address is, and thus can't know what type to use (p2pkh, p2sh-p2wpkh, p2wpkh), and thus has to assume all are possible (which results in at least labelling all 3). For `importmulti` we generally *do* have the address available. The goal here could be to provide a *default* address type (or explicitly name the type); not keep assuming it can be any.
* Relatedly, we don't have an encoding for "private key whose address is supposed to be P2SH-P2WPKH". My suggestion would be to add one (I believe Electrum has some sort of standard for this). `importprivkey` can't really use these, because it already has to assume it can be any type, but `importmulti` does not need to repeat this. It could assume just P2PKH for a legacy WIP encoding, and P2SH/P2WPKH when one of those novel encodings is used.
* `importprivkey` doesn't support key birth times; but this is not inherent, keys can have birth times.

**Comment by sipa** [comment]
> One question is if this should be simplified to where a user can put private keys in keys and leave scriptPubKey empty, or if it makes more sense to add a new field as I do here.

I think it should absolutely use `keys` - there is no need to add new fields to just provide the same functionality in a different way.

The way it should work IMO is that it integrates into all the functionality that exists (including labels, birth times, redeemscripts), but simply permit the scriptpubkey/address to be omitted and computed by  default from the private key - provided we have figured out how to deal with the encoding of different types of address derivation.

**Comment by kallewoof** [comment]
@sipa It seems Electrum 3.0 included an extension to WIF for segwit stuff: https://github.com/spesmilo/electrum/blob/5fef1e7980e6c9811448ad7d9fb6afa4460ac7fc/RELEASE-NOTES#L181-L194

A good starting point might be tol rework this PR to adopt that and see what people think?

File: src/wallet/rpcdump.cpp (hunk 3 above)
Comment (kallewoof) [reply src/wallet/rpcdump.cpp]: Why?

**Comment by kallewoof** [comment]
@promag Thanks for the feedback, see e896ba0, 6408244.

**Comment by kallewoof** [comment]
@sipa Small note:

> importprivkey doesn't support key birth times; but this is not inherent, keys can have birth times.

The method implementation does support birth times, in the `timestamp` argument which is always 1 for `importprivkey`, but uses the provided timestamp for `importmulti`.

File: src/wallet/rpcdump.cpp (hunk 3 above)
Comment (cdecker) [reply src/wallet/rpcdump.cpp]: Consistency?

**Comment by cdecker** [comment]
Concept ACK

Other than the quirks that `importprivkey` has to support mentioned by @sipa, which would be nice to avoid for a new call.

File: src/wallet/rpcdump.cpp (hunk 3 above)
Comment (kallewoof) [reply src/wallet/rpcdump.cpp]: Ahh, because of the other one below. Fixed.

**Comment by kallewoof** [comment]
Closing as too many aspects of this are up in the air.

**Pull request closed without merging**